### PR TITLE
DO NOT MERGE Nth attempt at migration test mock

### PR DESCRIPTION
**Pull request by notlesh**
This is my latest attempt at producing a test mock for `pallet_migrations`. I'm posting this here for discussion. It's a steaming pile of hot rust garbage.

## The problem

The original design of the Migrations pallet seems to work (I did a quick proof of concept where I plumbed an actual method invocation through a pallet, so I'm fairly confident in that).

Where things fall apart is in testing. It starts with the `Get` trait, which provides no context (other than potentially being generic over `Runtime`):

```rust
pub trait Get<T> {
    fn get() -> T;
}
```

This essentially requires that a `T` be produced from a static context. In this implementation, `T` is a Vec of `trait object`s:

```rust
type MigrationsList: Get<Vec<Box<dyn Migration>>>;
```

Where Migration is:

```rust
pub trait Migration {
    fn friendly_name(&self) -> &str;
    fn step(&self, previous_progress: Perbill, available_weight: Weight) -> (Perbill, Weight);
}
```

An example of a test case that I would like to support (using a `closure` capturing local stack variables by mutable reference):

```rust
#[test]
fn mock_migrations_static_hack_works() {
    let mut name_fn_called: bool = false;
    let mut step_fn_called: bool = false;

    // these closures map directly to the Migration trait fns
    something.registerCallback(
            &mut|| {
                name_fn_called = true;
                "hello, world"
            },
            &mut|_, _| -> (Perbill, Weight) {
                step_fn_called = true;
                (Perbill::zero(), 0u64.into())
            }
        );

    ExtBuilder::default().build().execute_with(|| {
        // this would:
        // 1) call our Get impl
        // 2) iterate over each trait object implementing Migration
        // 3) invoke the methods of each, effectively invoking our closures above
        Migrations::on_runtime_upgrade();
    });

    assert_eq!(name_fn_called, true, "mock migration should call friendly_name()");
    assert_eq!(step_fn_called, true, "mock migration should call step()");
}
```

There are a lot of pieces to this that complicated things, but the implied `'static` lifetime of using a singleton is the one that just doesn't play well with `FnMut`/closures. I don't think there's a way around this without bigger architectural changes (or maybe using `unsafe`).

Basically, the entire mock design mess boils down to this:
* I want a singleton to satisfy the need for the `Get` impl to provide a Vec with nothing but static context
* Each test would then create a list of closures (`FnMut`) that could be associated ("glued") with the Vec's contents
* When `pallet_migration` invokes a `Migration` callback, the singleton could look up the proper closure and invoke it

### CODE REVIEW

**Comment by librelois** [comment]
By default closures capture their environment by reference and not by ownership, so they can't impl Send or Sync.
If you need a closure to impl Send and Sync, you have to either:

1. That the closure captures its environment by becoming its owner -> add the `move` keyword in front of the closure.
2. Or the closure captures a reference to a static environment, but this only works without unsafe for Fn and FnOnce. Because FnMut would then capture a `&mut 'static T`, a reference that can only be used in unsafe context.

So indeed what you are trying to do can only be done with unsafe blocks, because here the borrow-checker is unable to statically guarantee the safety of the memory.

I had many problems with closures too, what unlocked me was to understand how they work internally: https://stackoverflow.com/questions/45935100/how-do-rust-closures-work-and-how-does-it-execute-a-closure

In your case: either you assume unsafe blocks by explaining in comments why it is safe, what dynamic controls make sure of it. Or you strongly re-architect ^^
